core/src/test/scala/unit/kafka/log/LogValidatorTest.scala

```diff
@@ -22,20 +22,43 @@ import java.util.concurrent.TimeUnit
 import kafka.api.{ApiVersion, KAFKA_2_0_IV1}
 import kafka.common.LongRef
 import kafka.message._
-import org.apache.kafka.common.errors.{InvalidTimestampException, UnsupportedCompressionTypeException, UnsupportedForMessageFormatException}
+import org.apache.kafka.common.errors.{InvalidTimestampException, KafkaStorageException, UnsupportedCompressionTypeException, UnsupportedForMessageFormatException}
 import org.apache.kafka.common.record._
 import org.apache.kafka.common.utils.Time
 import org.apache.kafka.test.TestUtils
 import org.junit.Assert._
 import org.junit.Test
-import org.scalatest.Assertions.intercept
+import org.scalatest.Assertions.{assertThrows, intercept}
 
 import scala.collection.JavaConverters._
 
 class LogValidatorTest {
 
   val time = Time.SYSTEM
 
+  @Test
+  def testOnlyOneBatchCompressedV0(): Unit = {
+    checkOnlyOneBatchCompressed(RecordBatch.MAGIC_VALUE_V0)
+  }
+
+  @Test
+  def testOnlyOneBatchCompressedV1(): Unit = {
+    checkOnlyOneBatchCompressed(RecordBatch.MAGIC_VALUE_V1)
+  }
+
+  @Test
+  def testOnlyOneBatchCompressedV2(): Unit = {
+    checkOnlyOneBatchCompressed(RecordBatch.MAGIC_VALUE_V2)
+  }
+
+  private def checkOnlyOneBatchCompressed(magic: Byte) {
+    LogValidator.validateCompressedRecords(createRecords(magic, 0L, CompressionType.GZIP))
```

Comment:
I wonder if it would be better to validate through `validateMessagesAndAssignOffsets` since that is the method which Log actually calls. Also, we may as well add a similar check to ensure that non-compressed v2 only accepts one batch.

Reply:
I thought about that too and was a bit lazy getting to pass in all those parameters, I can change that again.

Regarding the check for uncompressed V2: yeah I can do that too.